### ISSUE TITLE
[Bug] CLIPTokenizer: different sequence length and padding token inside tf.data vs standalone

### ISSUE THREAD

**Issue by yassienashrafwasfy** [keras-team/keras-hub#2730]
<html>
<body>
<html><head></head><body><h1>🐛 CLIPTokenizer produces inconsistent output inside <code>tf.data</code> vs standalone after PR #2628</h1>
<h2>Summary</h2>
<p>After commit <a href="https://github.com/keras-team/keras-hub/commit/e18a94fd2ca7784a13415f80f7b9b98ca8df6499"><code>e18a94f</code></a>, <code>CLIPTokenizer</code> produces <strong>different sequence lengths and padding tokens</strong> depending on whether it is called standalone (Python) or inside a <code>tf.data.map()</code> pipeline. This silently corrupts training data and causes shape mismatch errors at runtime.</p>

<h2>Root Cause</h2>
<p>PR <a href="https://github.com/keras-team/keras-hub/commit/e18a94fd2ca7784a13415f80f7b9b98ca8df6499">#2628 / commit e18a94f</a> introduced two separate code paths in <code>StartEndPacker.call()</code>:</p>
<pre><code class="language-python"># keras_hub/src/layers/preprocessing/start_end_packer.py
def call(self, ...):
    if not self._allow_python_workflow or in_tf_function():
        return self._call_tf(...)      # ← used inside tf.data.map
    else:
        return self._call_python(...)  # ← used standalone
</code></pre>
<p><a href="https://github.com/keras-team/keras-hub/commit/e18a94fd2ca7784a13415f80f7b9b98ca8df6499#diff-ddc6f978f8f822181cd1f805be69c2241317d1f19b0e2b39006cfa48b90903f0R1-R406">See the exact lines here</a></p>
<p>Since <code>CLIPTokenizer</code> internally uses <code>StartEndPacker</code>, it now routes through <strong>two completely different implementations</strong> depending on execution context — producing inconsistent results:</p>

Context | Code path | Output shape | Padding token
-- | -- | -- | --
Standalone Python | _call_python | (batch, 77) | 49407 ✅
Inside tf.data.map | _call_tf (via in_tf_function()) | (79,) | 0 ❌

<p>With <code>sequence_length=77</code>:</p>
<ul>
<li><strong>Standalone</strong>: outputs 77 tokens padded with <code>49407</code> (correct CLIP EOS token)</li>
<li><strong>Inside <code>tf.data</code></strong>: outputs 79 tokens padded with <code>0</code> (wrong length, wrong padding)</li>
</ul>
<hr>
<h2>Steps to Reproduce</h2>
<pre><code class="language-python">import numpy as np
import tensorflow as tf
import keras_hub

# Setup
tokenizer = keras_hub.models.CLIPTokenizer.from_preset(
    "clip_vit_large_patch14",
    sequence_length=77
)

def preprocessor(prompts):
    return tokenizer(prompts)

# ── Test 1: Standalone (Python mode) ──────────────────────────────
token_ids = preprocessor(["a photo of a dog"])
print(f"Standalone shape: {np.array(token_ids).shape}")

# ── Test 2: Inside tf.data.map ────────────────────────────────────
feature_description = {
    'caption': tf.io.FixedLenFeature([], tf.string),
}

def parse_example(serialized):
    example  = tf.io.parse_single_example(serialized, feature_description)
    caption  = example['caption']
    token_ids = preprocessor(caption)   # ← called with tf.string scalar
    return token_ids

ds = tf.data.TFRecordDataset(shards, compression_type='GZIP')
 ds = ds.map(parse_example, num_parallel_calls=tf.data.AUTOTUNE)

</code></pre>
<p><strong>Observed behavior:</strong></p>
<pre><code>Standalone:        shape=(1, 77),  padding_token=49407  ✅
Inside tf.data:    shape=(79,),    padding_token=0      ❌
</code></pre>
<p><strong>Expected behavior:</strong></p>
<p>Both contexts should produce identical output: <code>shape=(1, 77)</code> padded with <code>49407</code>.</p>

<h2>Workaround</h2>
<p>Force Python-mode tokenization inside <code>tf.data</code> using <code>tf.py_function</code>:</p>
<pre><code class="language-python">def parse_example(serialized):
    example  = tf.io.parse_single_example(serialized, feature_description)
    caption  = example['caption']

    token_ids = tf.py_function(
        func=lambda cap: preprocessor([cap.numpy().decode('utf-8')]),
        inp=[caption],
        Tout=tf.int32
    )
    token_ids.set_shape([77])
    return token_ids
</code></pre>

[Bug Reproduce](https://www.kaggle.com/code/yassienwasfy/clip-tokenizer-preprocessor-reproduce)
